Porting Subscription, Azure, Location, Region and NSG enhancements from Java SDK

martinsawicki opened the pull request:
Includes the following:
- `IAzure.GetCurrentSubscription()`
- `IAzure.Subscriptions`
- `Ilocation.Region`
- renaming `ISubscriptions.GetByName()` --> `.GetById()` and making it work
- fixing `ISubscription.ListLocations()`
- `ISubscription.GetLocationByRegion()`
- updated related tests and recordings

(Sorry, I keep forgetting to "git reset" until it's too late, so GitHub keep including commits from previously accepted PRs; this PR is only about commit  https://github.com/Azure/azure-sdk-for-net/commit/50eed19bafbffa0d804660dcebacaf89464adf77)

FYI - corresponding JAVA SDK commits:
- https://github.com/Azure/azure-sdk-for-java/commit/5fc0ed06a474fdfbef3767ee8e589deaf59d0a98
- https://github.com/Azure/azure-sdk-for-java/commit/91b04b119892bd26743d66af6ac4110dd30f3432
- https://github.com/Azure/azure-sdk-for-java/commit/d649ccd246e5cbabeafbe7de654ecfaa00ad3d10
- https://github.com/Azure/azure-sdk-for-java/commit/bbd2582e3149dcd6e0bdad8d461f86b96c1abdd2
- https://github.com/Azure/azure-sdk-for-java/commit/32b9e11b43e6ea2e2c9f3ae473880c1f17388e05
- https://github.com/Azure/azure-sdk-for-java/commit/24e6d7e7de754cea15a10c74402480bb86b63151

Commit:
correcting Region to use proper .NET enum naming/capitalization conventions

Commit:
Merge remote-tracking branch 'upstream/Fluent' into Fluent

# Conflicts:
#	Samples/ResourceManagement/AppService/ManageWebAppBasic/Program.cs
#	Samples/ResourceManagement/AppService/ManageWebAppSlots/Program.cs
#	Samples/ResourceManagement/AppService/ManageWebAppSourceControl/Program.cs
#	Samples/ResourceManagement/AppService/ManageWebAppSqlConnection/Program.cs
#	Samples/ResourceManagement/AppService/ManageWebAppStorageAccountConnection/Program.cs
#	Samples/ResourceManagement/AppService/ManageWebAppWithDomainSsl/Program.cs
#	Samples/ResourceManagement/AppService/ManageWebAppWithTrafficManager/Program.cs
#	Samples/ResourceManagement/Batch/ManageBatchAccount/Program.cs
#	Samples/ResourceManagement/Cdn/ManageCdn/Program.cs
#	Samples/ResourceManagement/Compute/CreateVMsUsingCustomImageOrSpecializedVHD/Program.cs
#	Samples/ResourceManagement/Compute/CreateVirtualMachinesInParallel/Program.cs
#	Samples/ResourceManagement/Compute/ListVirtualMachineExtensionImages/Program.cs
#	Samples/ResourceManagement/Compute/ListVirtualMachineImages/Program.cs
#	Samples/ResourceManagement/Compute/ManageAvailabilitySet/Program.cs
#	Samples/ResourceManagement/Compute/ManageVirtualMachine/Program.cs
#	Samples/ResourceManagement/Compute/ManageVirtualMachineExtension/Program.cs
#	Samples/ResourceManagement/Compute/ManageVirtualMachineScaleSet/Program.cs
#	Samples/ResourceManagement/Compute/ManageVirtualMachinesInParallel/Program.cs
#	Samples/ResourceManagement/Compute/ManageVirtualMachinesInParallelWithNetwork/Program.cs
#	Samples/ResourceManagement/Dns/ManageDns/Program.cs
#	Samples/ResourceManagement/KeyVault/ManageKeyVault/Program.cs
#	Samples/ResourceManagement/Network/ManageApplicationGateway/Program.cs
#	Samples/ResourceManagement/Network/ManageInternalLoadBalancer/Program.cs
#	Samples/ResourceManagement/Network/ManageInternetFacingLoadBalancer/Program.cs
#	Samples/ResourceManagement/Network/ManageIpAddress/Program.cs
#	Samples/ResourceManagement/Network/ManageNetworkInterface/Program.cs
#	Samples/ResourceManagement/Network/ManageNetworkSecurityGroup/Program.cs
#	Samples/ResourceManagement/Network/ManageSimpleApplicationGateway/Program.cs
#	Samples/ResourceManagement/Network/ManageVirtualNetwork/Program.cs
#	Samples/ResourceManagement/RedisCache/ManageRedis/Program.cs
#	Samples/ResourceManagement/ResourceManager/DeployUsingARMTemplate/Program.cs
#	Samples/ResourceManagement/ResourceManager/DeployUsingARMTemplateWithProgress/Program.cs
#	Samples/ResourceManagement/ResourceManager/ManageResource/Program.cs
#	Samples/ResourceManagement/ResourceManager/ManageResourceGroup/Program.cs
#	Samples/ResourceManagement/Sql/ManageSqlDatabase/Program.cs
#	Samples/ResourceManagement/Sql/ManageSqlDatabaseInElasticPool/Program.cs
#	Samples/ResourceManagement/Sql/ManageSqlDatabasesAcrossDifferentDataCenters/Program.cs
#	Samples/ResourceManagement/Sql/ManageSqlFirewallRules/Program.cs
#	Samples/ResourceManagement/Storage/ManageStorageAccount/Program.cs
#	Samples/ResourceManagement/TrafficManager/ManageTrafficManager/Program.cs

Commit:
renaming SharedSettings as SdkContext to be consistent with Java

Commit:
changing the capitalization/naming of enum constants to follow .NET conventions

Commit:
correcting misspelling in traffic manager namespace and deriving TM expandable enums from ExpandableStringEnum

Commit:
deriving all other single-string expandable enums from ExpandableStringEnum

Commit:
misc sample polish

Commit:
exposing .manager() consistently across networking-related collections
(equivalent of Java SDK commit: https://github.com/Azure/azure-sdk-for-java/commit/0314714eee50f6044e0a600f1d4fcfd70443743f)

Commit:
exposing .manager() consistently across app service-related resources
(equivalent to Java SDK checkin https://github.com/Azure/azure-sdk-for-java/commit/c9df856c5821acb71c522f7a2f2dd438b4d10fcd)

Commit:
exposing .manager() across batch-related collections
(equivalent to Java SDK checkin https://github.com/Azure/azure-sdk-for-java/commit/71d22358e365e14274d71d5810355a3b4c12b21f)

Commit:
exposing .manager() across CDN related collections
(equivalent to Java SDK checking https://github.com/Azure/azure-sdk-for-java/commit/ee20c6dbf8ed9e943803463ba85efb4b8ecf8e2e)

Commit:
exposing .manager() consistently across compute-related collections
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/484279b87a9b9d9c1c102b942a3862625dc23e1d)

Commit:
exposing .manager() consistently across DNS-related collections
(equivalent to Java SDK checkin https://github.com/Azure/azure-sdk-for-java/commit/c79faaaf128251649525c4beaac4cf47beda25f2)

Commit:
exposing .manager() consistently on Graph-related collections
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/9be59b823734c0c2fdb0b655db3606cb8618fd07)

Commit:
exposing .manager() on KeyVault-related collections
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/fc5681d6c7010b831e3e7da2a9a1ffdbb85eb172)

Commit:
exposing .manager() on Redis-related collections
(equivalent to Java SDK checkin https://github.com/Azure/azure-sdk-for-java/commit/092e61b2e85b29c6dd01f2bbea51b635a468a973)

Commit:
exposing .manager() across resource-related collections
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/a67248a01094b7880a6b3f76197977fc65a4092f)

Commit:
exposing .manager() on SQL related collections
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/45114972ed3a3832511c65819ed3221264ca9499)

Commit:
exposing .manager() on storage related collections
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/6e5ab809f8748cc172bef402d9c9fb686ea37b96)

Commit:
exposing .manager() on traffic manager related collections
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/57f0eafd0df197b7ba85d0527643fbf612dc4b2a)

Commit:
removing .setInner() from public access
(equivalent to Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/0add729abc81a706c6944e919ca08e8ec7f7ce77)

Commit:
Merge remote-tracking branch 'upstream/Fluent' into Fluent

# Conflicts:
#	Samples/ResourceManagement/Network/ManageApplicationGateway/Program.cs
#	Samples/ResourceManagement/Network/ManageInternalLoadBalancer/Program.cs
#	Samples/ResourceManagement/Network/ManageInternetFacingLoadBalancer/Program.cs
#	Samples/ResourceManagement/Network/ManageIpAddress/Program.cs
#	Samples/ResourceManagement/Network/ManageNetworkSecurityGroup/Program.cs
#	Samples/ResourceManagement/Network/ManageVirtualNetwork/Program.cs
#	src/ResourceManagement/AppService/Microsoft.Azure.Management.AppService.Fluent/JavaVersion.cs
#	src/ResourceManagement/AppService/Microsoft.Azure.Management.AppService.Fluent/PhpVersion.cs
#	src/ResourceManagement/AppService/Microsoft.Azure.Management.AppService.Fluent/PythonVersion.cs
#	src/ResourceManagement/AppService/Microsoft.Azure.Management.AppService.Fluent/WebContainer.cs
#	src/ResourceManagement/Azure.Fluent/Fluent.Tests/Common/ApplicationGatewayHelper.cs
#	src/ResourceManagement/Azure.Fluent/Fluent.Tests/Common/NetworkTestHelperBase.cs
#	src/ResourceManagement/Compute/Microsoft.Azure.Management.Compute.Fluent/ComputeUsageUnit.cs
#	src/ResourceManagement/Compute/Microsoft.Azure.Management.Compute.Fluent/PowerState.cs
#	src/ResourceManagement/TrafficManager/Microsoft.Azure.Management.TrafficManager.Fluent/ProfileDnsNameUnavailableReason.cs

Commit:
porting Subscription, Azure, Location and Region enhancements from Java SDK

Commit:
Merge remote-tracking branch 'upstream/Fluent' into Fluent

Commit:
porting ISubscription.GetLocationByRegion() from the Java side

azurecla commented:
Hi __@martinsawicki__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (marcins). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

Commit:
renaming LocationInner back to Location, to prevent breakage when regenerated, but we should fix the generator

Commit:
enabling NSG to be deleted cleanly, even if NIC refs exist
(port of Java SDK commit https://github.com/Azure/azure-sdk-for-java/commit/24e6d7e7de754cea15a10c74402480bb86b63151)

Commit:
updated recording of network test

hovsepm commented:
@martinsawicki merged. Please reset your branch ;)
